runtimes/README.md

```diff
@@ -225,10 +225,12 @@ Complete Identity Management parameter and result interfaces can be found in [id
 
 ### Notification
 
-The notification feature can be used to send custom customer-facing notifications to clients. Notifications can contain actions, like show URL, but also followup actions, like request customer acknowledgement. When customer reacts to followup actions, asyncronous notification can be sent from client to server to notify server about this.
+The notification feature can be used to send custom customer-facing notifications to clients. Notifications can contain actions, like show URL, but also followup actions, like request customer acknowledgement. When customer reacts to followup actions, asynchronous notification can be sent from client to server to notify server about this.
 
 Notifications should be used in rare / exceptional cases that require customer attention (like some change happened or action recommended), but are not blocking the main flow. Clients can decide to throttle notifications, if too many are sent.
 
+Consider using LSP [ShowMessage notification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessage) instead, if your notification does not require actions or followup actions.
```

Comment:
LSP's ShowMessageRequest allows for MessageActionItems which effectively achieve the same thing.  I'm struggling to think of how this differs and questioning whether ShowMessageRequest is the correct approach.  

The one thing I noticed about ShowMessageRequest is that it appears to be for modal use based on the description in the spec and the fact that is a request that returns the selected action in its response.  While JSON-RPC and LSP state that requests don't need to be completed before sending other requests and notifications, there isn't a recommendation about using requests like long-polling or leaving them open for long periods of time.

The notifications for toolkit (and presumably for implementors) are modeless/asynchronous in nature regarding if and when the user selects an action, so ShowMessageRequest would not be suitable for that use.

What toolkits needs is in the ShowMessage space, but blurs the lines between the ShowMessage notification and the ShowMessage request in that it's fire and forget like the ShowMessage notification, but allows for actions to be supplied like the ShowMessage request.

The combination of ShowMessage notification and ExecuteCommand almost achieve what is needed in the Notifications API.  The missing piece is that there isn't a way in the LSP spec to provide commands via ShowMessage notification.  I also don't see a generic way of providing commands from a server to a client in any other request/notifications.  If there was a ShowMessageModeless notification that passed an array of Commands in its params, that would be what is needed for the Notifications API I think.

I wonder if Larry considered this or an approach more aligned with LSP when designing?

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessage

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessageRequest

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#command

Reply:
Why did you make conclusion it's for modal use? They are not mentioning this explicitly. Unfortunately, they also do not tell much about long waiting requests or any timeouts - maybe there are none. There are some short mentions of possibility of reordering. I think the best way to figure out if it actually works would be to test in some test clients. However, I'm not sure if different IDEs can have different behavior.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#messageOrdering

Reply:
I guess main challenge with using standard `ShowMessage` is to be sure that all our custom clients (Toolkits) are able to intercept that message and wrap it with needed UI. I imagine there might be situations, when IDE Client does not allow that and those destinations will have degraded/not optimal UX for custom notifications handling

Reply:
Also another concern is that it maybe not possible to extend standard LSP Request params for more complex follow-up actions, required by other features